src/index.js

```diff
@@ -19,7 +19,27 @@ plugins.flow = flowPlugin;
 plugins.jsx = jsxPlugin;
 
 export function parse(input, options) {
-  return new Parser(options, input).parse();
+  if (options && options.sourceType === "unambiguous") {
+    options = Object.assign({}, options);
+    try {
+      options.sourceType = "module";
+      const ast = new Parser(options, input).parse();
+
+      // Rather than try to parse as a script first, we opt to parse as a module and convert back
+      // to a script where possible to avoid having to do a full re-parse of the input content.
+      if (!hasModuleSyntax(ast)) ast.program.sourceType = "script";
```

Comment:
The "correct logic" would be to parse as script first and module later if that fails, but true, it's a lot more likely to get code that parses as both (or module only) as input. Not a lot of people still have code that _only_ parses in sloppy mode.